### PR TITLE
Improve exception message of function regexp_extract

### LINKED ISSUE #56393: Poor usability of regexpExtract
See https://play.clickhouse.com/play?user=play#c2VsZWN0IHJlZ2V4cEV4dHJhY3QoYm9keSwgJycpIGFzIGxpbmssICogZnJvbSBnaXRodWJfZXZlbnRz

```
select regexpExtract(body, '') as link, * from github_events

"exception": "Code: 353. DB::Exception: Index value 1 is out of range, should be in [0, 1). (INDEX_OF_POSITIONAL_ARGUMENT_IS_OUT_OF_RANGE) (version 23.10.1.1954 (official build))"
```

### PR DESCRIPTION
**Pull request by taiyang-li**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improve exception message of function regexp_extract, close https://github.com/ClickHouse/ClickHouse/issues/56393

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit 7aea3847dbea8a4d25b9aaafff5250edfbb47f1e with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59319/7aea3847dbea8a4d25b9aaafff5250edfbb47f1e/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/59319/7aea3847dbea8a4d25b9aaafff5250edfbb47f1e/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by taiyang-li** [comment]
@tavplubix  need you review, thanks!

File: src/Functions/regexpExtract.cpp [hunk 1]
```diff
@@ -124,21 +124,23 @@ class FunctionRegexpExtract : public IFunction
...
-        ColumnString::Offsets & res_offsets)
+        ColumnString::Offsets & res_offsets) const
     {
         const OptimizedRegularExpression regexp = Regexps::createRegexp<false, false, false>(pattern);
         unsigned capture = regexp.getNumberOfSubpatterns();
         if (index < 0 || index >= capture + 1)
             throw Exception(
                 ErrorCodes::INDEX_OF_POSITIONAL_ARGUMENT_IS_OUT_OF_RANGE,
-                "Index value {} is out of range, should be in [0, {})",
+                "Index value {} is out of range of regexp pattern '{}' in function {}, should be in [0, {})",
```
Comment (rschu1ze) [inline src/Functions/regexpExtract.cpp:140]: The error message now sounds slightly weird. Maybe

```
Index value {} for regexp pattern `{}` in function {} is out-of-range, should be in [0, {})`?
```

Comment (taiyang-li) [reply src/Functions/regexpExtract.cpp:140]: Sounds better. I'll fix it.

**Pull request merged**